### PR TITLE
nixos/podman-dnsname: init

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Allow containers to contact each other by name.
Name resolution in container networks is rather useful.

This makes podman a near-feature complete alternative to docker, unblocking arion users who need systemd in their containers.

###### Notes

The test is rather small, so it makes sense to merge it into the `podman` test, but that file is changing with #123841

cc @NixOS/podman 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124589` at 0da67424 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>cniPlugins.dnsname</li>
    <li>nixos-install-tools</li>
  </ul>
</details>

**Comment by zowoq** [comment]
I'm going to have several comments for this but I'm not going to have any time in the next few days, sorry.

**Comment by roberth** [comment]
> I'm going to have several comments for this but I'm not going to have any time in the next few days, sorry.

I understand and I'm really grateful for your reviews on my earlier pr #123841. For me, having to switch from docker to podman was unexpected so I had to make time (and I hated that). I don't expect you to do the same.

File: nixos/tests/podman-dnsname.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import ./make-test-python.nix (
```
Comment (roberth) [inline nixos/tests/podman-dnsname.nix:1]: As mentioned, I'll happily merge this into the `podman` test after #123841 (unless you want it to be separate)

**Review by roberth — COMMENTED** [review]
I've self-reviewed to highlight some of the decisio possible issues I've made. I hope that what remains will be trivial.

File: nixos/modules/virtualisation/podman-dnsname.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    mkIf
+    types
+    ;
+
+  cfg = config.virtualisation.podman.dnsname;
+
+in
+{
+  options = {
+    virtualisation.podman.dnsname = {
```
Comment (zowoq) [inline nixos/modules/virtualisation/podman-dnsname.nix:14]: I'm not necessarily :-1:  on this but it doesn't seem quite right? By itself It's an option to add the `dnsname` plugin path to `containersConf` which we'll also have (manually) via `containersConf.cniPlugins`?

Could we do `virtualisation.podman.defaultNetwork.dnsname.enable` which adds the config and sets the plugin path?

I think that may also better align the options with your comment about a possible future rename?

File: nixos/modules/virtualisation/podman.nix [hunk 3]
```diff
@@ -22,9 +23,23 @@ let
...
   '';
 
+  net-conflist = pkgs.runCommand "87-podman-bridge.conflist" {
+    nativeBuildInputs = [ pkgs.jq ];
+    extraPlugins = builtins.toJSON cfg.defaultNetwork.extraPlugins;
+    jqScript = ''
+      . + { "plugins": (.plugins + $extraPlugins) }
+    '';
+  } ''
+    jq <${pkgs.podman-unwrapped.src}/cni/87-podman-bridge.conflist \
```
[the remark this answers is not in the document]
Comment (roberth) [reply nixos/modules/virtualisation/podman.nix:33]: The source derivation wasn't part of the output closure, but it's a good change nonetheless. With your pr, NixOS users won't have to download the podman sources for this one file.

File: nixos/modules/virtualisation/podman-dnsname.nix (hunk 2 above)
Comment (roberth) [reply nixos/modules/virtualisation/podman-dnsname.nix]: Done.

**Comment by roberth** [comment]
It's ready for review again.

Turns out I can't read; `dnsname-cni` package already existed!

**Comment by zowoq** [comment]
This looks good, thanks!

**Pull request merged**